pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+	for d.processNextItem() {
+		// continue looping
+	}
+}
+
+func (d *ServiceDNSEndpointController) processNextItem() bool {
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	err := d.processItem(key.(string))
+
+	if err == nil {
+		// No error, tell the queue to stop tracking history
+		d.queue.Forget(key)
+	} else if d.queue.NumRequeues(key) < maxRetries {
+		glog.Errorf("Error processing %s (will retry): %v", key, err)
+		// requeue the item to work on later
+		d.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		glog.Errorf("Error processing %s (giving up): %v", key, err)
+		d.queue.Forget(key)
+		runtime.HandleError(err)
+	}
+
+	return true
+}
+
+func (d *ServiceDNSEndpointController) processItem(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Processing change to ServiceDNSEndpoint %s", key)
+	defer func() {
+		glog.V(4).Infof("Finished processing ServiceDNSEndpoint %q (%v)", key, time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := d.serviceDNSObjectStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+	}
+
+	if !exists {
+		//delete corresponding DNSEndpoint object
+		return d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Delete(name, &metav1.DeleteOptions{})
+	}
+
+	dnsObject, ok := obj.(*feddnsv1a1.MultiClusterServiceDNSRecord)
+	if !ok {
+		return fmt.Errorf("recieved event for unknown object %v", obj)
+	}
+
+	dnsEndpoints, err := GetEndpointsForServiceDNSObject(dnsObject)
+	if err != nil {
+		return err
+	}
+
+	dnsEndpointObject, err := d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			newDNSEndpointObject := &feddnsv1a1.DNSEndpoint{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: namespace,
+				},
+				Spec: feddnsv1a1.DNSEndpointSpec{
+					Endpoints: dnsEndpoints,
+				},
+			}
+
+			_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Create(newDNSEndpointObject)
+		}
+		return err
+	}
+
+	dnsEndpointObject.Spec.Endpoints = dnsEndpoints
+	_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Update(dnsEndpointObject)
+	return err
+}
+
+// GetEndpointsForServiceDNSObject returns endpoint objects for each MultiClusterServiceDNSRecord object that should be processed.
+func GetEndpointsForServiceDNSObject(dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) ([]*feddnsv1a1.Endpoint, error) {
+	endpoints, err := getNamesForDNSObject(dnsObject, dnsObject.Spec.FederationName, dnsObject.Spec.DNSSuffix)
+	if err != nil {
+		return nil, err
+	}
+
+	return DedupeAndMergeEndpoints(endpoints), nil
+}
+
+// Merge and remove duplicate endpoints
+func DedupeAndMergeEndpoints(endpoints []*feddnsv1a1.Endpoint) (result []*feddnsv1a1.Endpoint) {
```

Comment:
(No action required) Consider making this function package local since it is not used outside of the package.